### PR TITLE
Feature/add cantabular edit metadata screen

### PR DESCRIPTION
**Pull request by patrickjfl**
### What

We have added a new 'Edit metadata' screen for handling metadata for Cantabular datasets, with a feature flag.

### How to review

Read the code, run the test suite, post a recipe and click through the Cantabular dataset journey.
The enableCantabularJourney flag is by default set to false, to set it to true, run `export ENABLE_CANTABULAR_JOURNEY=true` before running `make debug`.

### Who can review

Anyone but me or Valentina

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -151,6 +152,12 @@ const Index = () => {
                                                 component={userIsAuthenticated(EditMetadataItem)}
                                             />
                                         </Route>
+                                        <Route path={`versions/:versionID/cantabular`} component={config.enableCantabularJourney ? userIsAuthenticated(CantabularMetadataController) : null}>
```
Comment (rafahop) [inline src/index.js:155]: Is it worth having the flag here? If it is off we won't access this route and because this is only an internal system I wouldn't control this with the feature flag so to reduce the changes we'll need to make once we remove it? But there is nothing wrong with this if you prefer to keep it.

File: src/app/components/simple-editable-list/SimpleEditableList.jsx [hunk 2]
```diff
@@ -59,7 +60,7 @@ export default class SimpleEditableList extends Component {
                                     field={field}
                                     handleEditClick={this.handleEditClick}
                                     handleDeleteClick={this.handleDeleteClick}
-                                    disabled={this.props.disableActions}
+                                    disabled={this.props.disableActions || this.props.disableEditFunctionality}
```
Comment (rafahop) [inline src/app/components/simple-editable-list/SimpleEditableList.jsx:63]: I see we always use `disableEditFunctionality` together with `disableActions`, could we reuse the existing prop instead?

File: src/index.js (hunk 1 above)
Comment (vmilitaru) [reply src/index.js:155]: I've removed the flag to minimise the changes.

File: src/app/components/simple-editable-list/SimpleEditableList.jsx (hunk 2 above)
Comment (vmilitaru) [reply src/app/components/simple-editable-list/SimpleEditableList.jsx]: I've also taken out the `disableEditFunctionality` and update the `disableActions` prop so that the component is disabled if there is any metadata coming through for that particular field.

File: src/app/views/workflow-preview/WorkflowPreview.jsx [hunk 3]
```diff
@@ -21,11 +22,25 @@ const propTypes = {
...
-    handleBackButton = () => {
+    handleBackButton = async () => {
         const previousUrl = url.resolve("../");
-        this.props.dispatch(push(previousUrl));
+        await this.getDatasetType(this.props.params.datasetID);
+        if (this.state.cantabularDataset) {
+            this.props.dispatch(push(previousUrl + "/cantabular"));
+        } else {
+            this.props.dispatch(push(previousUrl));
+        }
```
Comment (gilesburdett) [inline src/app/views/workflow-preview/WorkflowPreview.jsx:35]: Lines 31 to 35 can be replaced by:

```
this.props.dispatch(push(this.state.cantabularDataset ? `${previousUrl}/cantabular` : previousUrl));
```

But that's a bit gnarly, so you could switch things around and do this:

```
handleBackButton = async () => {
  await this.getDatasetType(this.props.params.datasetID);
  const previousUrl = `${url.resolve("../")}${this.state.cantabularDataset ? /cantabular : ""}`
  this.props.dispatch(push(previousUrl));
}
```

File: src/app/views/workflow-preview/WorkflowPreview.jsx [hunk 4]
```diff
@@ -21,11 +22,25 @@ const propTypes = {
...
+    };
+
+    getDatasetType = datasetID => {
+        return datasets.get(datasetID).then(response => {
+            if (response.next.type === "cantabular_table" || response.next.type === "cantabular_flexible_table") {
+                this.setState({ cantabularDataset: true });
```
Comment (gilesburdett) [inline src/app/views/workflow-preview/WorkflowPreview.jsx:41]: This doesn't allow you to falsify the flag, so perhaps this would be better:

```
const type = response.next.type;
this.setState({ cantabularDataset: type === "cantabular_table" || type === "cantabular_flexible_table" });
```

File: src/app/views/datasets-new/versions/DatasetVersionsController.jsx [hunk 5]
```diff
@@ -28,16 +30,29 @@ export class DatasetVersionsController extends Component {
...
         const datasetID = this.props.params.datasetID;
         const editionID = this.props.params.editionID;
-
+        if (this.props.enableCantabularJourney) {
+            this.getDatasetType(datasetID);
+        }
         this.getAllVersions(datasetID, editionID);
     }
 
+    getDatasetType = datasetID => {
```
Comment (gilesburdett) [inline src/app/views/datasets-new/versions/DatasetVersionsController.jsx:46]:
```
getDatasetType = async datasetID => {
  await response = datasets.get(datasetID);
  const type = response.next.type;
  this.setState({ cantabularDataset: type === "cantabular_table" || type === "cantabular_flexible_table" });
}
```

File: src/app/views/datasets-new/versions/DatasetVersionsController.jsx [hunk 6]
```diff
@@ -117,7 +132,9 @@ export class DatasetVersionsController extends Component {
                 return {
                     id: version.id,
                     title: version.title,
-                    url: this.props.location.pathname + "/versions/" + version.version,
+                    url: this.state.cantabularDataset
```
Comment (gilesburdett) [inline src/app/views/datasets-new/versions/DatasetVersionsController.jsx:135]:
```
url: `${this.props.location.pathname}/versions/${version.version}${this.state.cantabularDataset ? "/cantabular" : ""}`
```

But better as:

```
const path = `${this.props.location.pathname}/versions/${version.version}`
return {
   ...
   url: `${path}${this.state.cantabularDataset ? "/cantabular" : ""}`
   ...
};
```

File: src/app/views/datasets-new/edit-metadata-cantabular/CantabularMetadataController.jsx [hunk 7]
```diff
@@ -0,0 +1,755 @@
...
+                notifications.add({
+                    type: "warning",
+                    message: `An error occured when getting information about this dataset version metadata. Please try refresh the page`,
+                    isDismissable: true,
+                });
+                console.error("get metadata: error GETting dataset metadata from controller", error);
+            });
+    };
+
+    getCantMetadata = (datasetID, nonCantDatasetMetadata) => {
```
Comment (gilesburdett) [inline src/app/views/datasets-new/edit-metadata-cantabular/CantabularMetadataController.jsx:126]: "Cant" rather unfortunately reads like "Can't" as in "Can not", so it would be better to rename this to `getCantabularMetadata` rather than saving a few key strokes.

File: src/app/views/datasets-new/edit-metadata-cantabular/CantabularMetadataController.jsx [hunk 8]
```diff
@@ -0,0 +1,755 @@
...
+        return datasets
+            .get(datasetID)
+            .then(dataset => {
+                const cantDatasetId = dataset.next.is_based_on["@id"];
```
Comment (gilesburdett) [inline src/app/views/datasets-new/edit-metadata-cantabular/CantabularMetadataController.jsx:130]: A similar issue here with "cant"

File: src/app/views/datasets-new/edit-metadata-cantabular/CantabularMetadataController.jsx [hunk 9]
```diff
@@ -0,0 +1,755 @@
...
+        const mapped = this.mapMetadataToState(nonCantDatasetMetadata, cantabularMetadata);
+        if (mapped.state === "associated" && mapped.collection !== this.props.params.collectionID) {
+            this.setState({ disableScreen: true });
+            notifications.add({
+                type: "neutral",
+                message: `This dataset is in another collection.`,
+                isDismissable: true,
+            });
+        }
+        mapped.state === "created"
```
Comment (gilesburdett) [inline src/app/views/datasets-new/edit-metadata-cantabular/CantabularMetadataController.jsx:285]: How about:

```
const isCreated = mapped.state === "created";
this.setState({ allowPreview: !isCreated, disableCancel: !isCreated });
```

File: src/app/views/datasets-new/edit-metadata-cantabular/CantabularMetadataController.jsx [hunk 10]
```diff
@@ -0,0 +1,755 @@
...
+            datasetState: mapped.state,
+            lastEditedBy: mapped.lastEditedBy,
+            versionIsPublished: mapped.versionIsPublished,
+        });
+    };
+
+    mapRelatedContentToState = (relatedDatasets, datasetID) => {
+        try {
+            return relatedDatasets.map((link, index) => {
+                return {
```
Comment (gilesburdett) [inline src/app/views/datasets-new/edit-metadata-cantabular/CantabularMetadataController.jsx:302]: There is a shorthand for returning an object that's useful when that's all a function has to do. So instead of

```
map((foo, bar) => {
  return {
    foo,
    bar
  };
})
```

you can wrap the object in parentheses and omit the `return` which is now implicit, like this:

```
map((foo, bar) => ({
  foo,
  bar
}))
```

File: src/app/views/datasets-new/edit-metadata-cantabular/CantabularMetadataController.jsx [hunk 11]
```diff
@@ -0,0 +1,755 @@
...
+        });
+        const newMetadataState = {
+            ...this.state.metadata,
+            dimensions: newDimensionMetadata,
+        };
+        this.setState({ metadata: newMetadataState, dimensionsUpdated: true });
+    };
+
+    handleSimpleEditableListAdd = stateFieldName => {
+        this.props.dispatch(push(`${this.props.location.pathname}/edit/${stateFieldName}/${this.state.metadata[stateFieldName].length}`));
```
Comment (gilesburdett) [inline src/app/views/datasets-new/edit-metadata-cantabular/CantabularMetadataController.jsx:448]: Could break this up into two lines, one `const path = <big long string>` and `this.props.dispatch(push(path))`

File: src/app/views/datasets-new/edit-metadata-cantabular/CantabularMetadataController.jsx [hunk 12]
```diff
@@ -0,0 +1,755 @@
...
+                lastest_changes: this.state.metadata.latestChanges,
+            },
+            dimensions: [...this.state.metadata.dimensions],
+            collection_id: this.props.params.collectionID,
+            collection_state: this.mapCollectionState(isSubmittingForReview, isMarkingAsReviewed),
+        };
+    };
+
+    mapCollectionState = (isSubmittingForReview, isMarkingAsReviewed) => {
+        const StatusInProgress = "InProgress";
```
Comment (gilesburdett) [inline src/app/views/datasets-new/edit-metadata-cantabular/CantabularMetadataController.jsx:615]: This may be inherited from the original in which case this remark is out of scope for this PR so please ignore it if that's the case, but ideally shouldn't states like this be defined in a single place for use throughout the application rather than defined locally and sent out into the application for consumption?

**Review by gilesburdett — CHANGES_REQUESTED** [review]

File: src/app/views/collections/details/CollectionDetailsController.jsx [hunk 13]
```diff
@@ -350,9 +352,14 @@ export class CollectionDetailsController extends Component {
...
         if (page.type === "dataset_version") {
-            const newURL = url.resolve(
+            let newURL = url.resolve(
                 `/collections/${this.props.activeCollection.id}/datasets/${page.datasetID}/editions/${page.edition}/versions/${page.version}`
             );
+            if (this.state.isCantabularDataset) {
+                newURL = url.resolve(
+                    `/collections/${this.props.activeCollection.id}/datasets/${page.datasetID}/editions/${page.edition}/versions/${page.version}/cantabular`
+                );
+            }
```
[the remark this answers is not in the document]
Comment (vmilitaru) [reply src/app/views/collections/details/CollectionDetailsController.jsx:362]: Done :white_check_mark:

File: src/app/views/collections/details/CollectionDetailsController.jsx [hunk 14]
```diff
@@ -588,6 +595,20 @@ export class CollectionDetailsController extends Component {
         this.handleRestoreDeletedContentClose();
     };
 
+    getDatasetType = pageID => {
+        if (pageID && pageID.includes("/")) {
+            let pageIDArray = pageID.split("/");
+            const datasetID = pageIDArray[0].trim();
```
[the remark this answers is not in the document]
Comment (vmilitaru) [reply src/app/views/collections/details/CollectionDetailsController.jsx:601]: Done :white_check_mark:

File: src/app/views/collections/details/CollectionDetailsController.jsx [hunk 15]
```diff
@@ -588,6 +595,20 @@ export class CollectionDetailsController extends Component {
...
+            return datasets.get(datasetID).then(response => {
+                if (response.next.type === "cantabular_flexible_table" || response.next.type === "cantabular_table") {
+                    this.setState({ isCantabularDataset: true });
+                } else {
+                    this.setState({ isCantabularDataset: false });
+                }
+            });
+        }
+    };
```
[the remark this answers is not in the document]
Comment (vmilitaru) [reply src/app/views/collections/details/CollectionDetailsController.jsx:610]: Nice refactoring  - done :white_check_mark:

File: src/app/views/workflow-preview/WorkflowPreview.jsx (hunk 3 above)
Comment (vmilitaru) [reply src/app/views/workflow-preview/WorkflowPreview.jsx]: Done :white_check_mark:

File: src/app/views/workflow-preview/WorkflowPreview.jsx (hunk 4 above)
Comment (vmilitaru) [reply src/app/views/workflow-preview/WorkflowPreview.jsx]: Done :white_check_mark:

File: src/app/views/datasets-new/versions/DatasetVersionsController.jsx (hunk 6 above)
Comment (vmilitaru) [reply src/app/views/datasets-new/versions/DatasetVersionsController.jsx:135]: Done :white_check_mark:

File: src/app/views/datasets-new/versions/DatasetVersionsController.jsx (hunk 5 above)
Comment (vmilitaru) [reply src/app/views/datasets-new/versions/DatasetVersionsController.jsx:46]: Done :white_check_mark:

File: src/app/views/datasets-new/edit-metadata-cantabular/CantabularMetadataController.jsx (hunk 9 above)
Comment (vmilitaru) [reply src/app/views/datasets-new/edit-metadata-cantabular/CantabularMetadataController.jsx]: Done :white_check_mark:

File: src/app/views/datasets-new/edit-metadata-cantabular/CantabularMetadataController.jsx (hunk 10 above)
Comment (vmilitaru) [reply src/app/views/datasets-new/edit-metadata-cantabular/CantabularMetadataController.jsx]: Done :white_check_mark:

File: src/app/views/datasets-new/edit-metadata-cantabular/CantabularMetadataController.jsx (hunk 11 above)
Comment (vmilitaru) [reply src/app/views/datasets-new/edit-metadata-cantabular/CantabularMetadataController.jsx]: Done :white_check_mark:

**Review by gilesburdett — APPROVED** [review]
